src/main/java/de/tum/in/www1/artemis/service/QuizSubmissionService.java

```diff
@@ -136,48 +136,58 @@ public QuizSubmission saveSubmissionForLiveMode(Long exerciseId, QuizSubmission
         if (!quizExercise.isQuizStarted() || quizExercise.isQuizEnded()) {
             throw new QuizSubmissionException("The quiz is not active");
         }
 
-        var batch = quizBatchService.getQuizBatchForStudent(quizExercise, user);
+        var batch = quizBatchService.getQuizBatchForStudentByLogin(quizExercise, userLogin);
         if (batch.stream().noneMatch(QuizBatch::isSubmissionAllowed)) {
             throw new QuizSubmissionException("The quiz is not active");
         }
 
-        var cachedSubmission = quizScheduleService.getQuizSubmission(exerciseId, user.getLogin());
+        var cachedSubmission = quizScheduleService.getQuizSubmission(exerciseId, userLogin);
         if (cachedSubmission.isSubmitted()) {
-            // current attempt has not yet been committed to DB, so countByExerciseIdAndStudentId will not pick it up
+            // the old submission has not yet been processed, so don't allow a new one yet
             throw new QuizSubmissionException("You have already submitted the quiz");
         }
 
-        // TODO: add one additional check: fetch quizSubmission.getId() with the corresponding participation and check that the user of participation is the
-        // same as the user who executes this call. This prevents injecting submissions to other users
-
-        // check if user has attempts left for this quiz
-        int submissionCount = submissionRepository.countByExerciseIdAndStudentId(exerciseId, user.getId());
-        log.debug("{} Counted {} submissions for user {} in quiz {} in {} µs.", logText, submissionCount, user.getLogin(), exerciseId, (System.nanoTime() - start) / 1000);
-        if (quizExercise.getAllowedNumberOfAttempts() != null && submissionCount >= quizExercise.getAllowedNumberOfAttempts()) {
-            throw new QuizSubmissionException("You have no more attempts at this quiz left");
+        if (quizExercise.getQuizMode() == QuizMode.SYNCHRONIZED) {
+            // in synchronized mode we cache the participation after we processed the submission, so we can check there if the submission was already processed
+            var cachedParticipation = quizScheduleService.getParticipation(exerciseId, userLogin);
+            if (cachedParticipation != null && cachedParticipation.getResults().stream().anyMatch(r -> r.getSubmission().isSubmitted())) {
+                throw new QuizSubmissionException("You have already submitted the quiz");
+            }
+        }
+        else {
+            // in other modes we have to check the db if there has already been a submission
+            int submissionCount = submissionRepository.countByExerciseIdAndStudentLogin(exerciseId, userLogin);
+            log.debug("{} Counted {} submissions for user {} in quiz {} in {} µs.", logText, submissionCount, userLogin, exerciseId, (System.nanoTime() - start) / 1000);
```

Comment:
Should this stay in the code?

Reply:
while thinking about the review above i removed that entirely query as well, so the logging is gone now as well.